Homework 6

JeRusakow opened the pull request.

JeRusakow committed:
set up for 6th homework

JeRusakow committed:
Merge branch 'main' of https://github.com/JeRusakow/epam_python_course into hw6

JeRusakow committed:
Task1 solved

JeRusakow committed:
Task2 solved

RomanTsaregorodtsev commented on homework6/task1/counter.py:
```diff
@@ -0,0 +1,53 @@
+    reset_instance_counter # resets the instance counter and returns the amount before resetting
+    """
+
+    original_init = getattr(cls, "__init__")
```
А чем просто `cls.__init__` тебя не устроил?

JeRusakow replied:
Действительно, одно и то же. Писал просто первое, что в голову пришло
`cls.__init__` выглядит проще, исправил

RomanTsaregorodtsev commented on homework6/task1/counter.py:
```diff
@@ -0,0 +1,53 @@
+
+    original_init = getattr(cls, "__init__")
+
+    setattr(cls, "instance_counter", 0)
```
`cls.instance_counter = 0`? 
+ я бы этот атрибут инкапсулировал, потому что мы делаем доступ к нему через get и reset

JeRusakow replied:
инкапсулировал

RomanTsaregorodtsev commented on homework6/task1/counter.py:
```diff
@@ -0,0 +1,53 @@
+        cls.instance_counter += 1
+        return original_init(self, *args, **kwargs)
+
+    def get_created_instances(self=None):
```
По сути это методы класса, а значит их и нужно сделать методами класса со всеми вытекающими

RomanTsaregorodtsev commented on homework6/task1/counter.py:
```diff
@@ -0,0 +1,53 @@
+    def get_created_instances(self=None):
+        return cls.instance_counter
+
+    def reset_instances_counter(self=None):
```
Тоже метод класса

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+from collections import defaultdict
+
+
+class DeadlineError(Exception):
```
Чёт докстринги потерялись. Нужно добавить как докстринги классов, так и докстринги методов.
Как именно можно почитать [здесь](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html): я предпочитаю google docstring

JeRusakow replied:
Написал документацию

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+    def do_homework(self, homework: Homework, solution: str):
+        if homework.is_active():
+            return HomeworkResult(self, homework, solution)
+        else:
```
Лишняя вложенность. else здесь не нужен

JeRusakow replied:
Исправил

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+
+
+class Student(Human):
+    def do_homework(self, homework: Homework, solution: str):
```
Типизация возвращаемого значения

JeRusakow replied:
Имеется два определения:
```
class Student(Human):
    def do_homework(self, homework: Homework, solution: str) -> HomeworkResult:
```
и
```
class HomeworkResult:
    def __init__(self, author: Student, task: Homework, solution: str):
```
Чтобы типизация всюду работала, классам нужно наперёд знать о существовании друг друга. Разнести в разные модули тоже не получится из-за ошибки циклических импортов. Решить это можно аналогом  `forward` из C++, но, насколько, почитал, в питоне такого нету. Как быть?

JeRusakow replied:
Разобрался

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+    homework_done = defaultdict(list)
+
+    @staticmethod
+    def create_homework(text: str, days_to_complete: int):
```
Типизация

JeRusakow replied:
Исправил

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+        if len(done_homework.solution) > 5:
+            Teacher.homework_done[done_homework.task].append(done_homework)
+            return True
+        else:
```
`else` тут не нужен

JeRusakow replied:
Исправил

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+    def create_homework(text: str, days_to_complete: int):
+        return Homework(text, days_to_complete)
+
+    def check_homework(self, done_homework: HomeworkResult) -> bool:
```
Какой IDE ты пользуешься? По идее она должна тебе подчёркивать, что ты не использовал `self` в методе, хотя определил его как метод объекта. Но вместо этого ты использовал `Teacher`. Чуешь к чему клоню?

RomanTsaregorodtsev replied:
Плюс плохо использовать имя класса внутри определения самого класса. Вот почему:
[one](https://stackoverflow.com/questions/46128877/should-i-use-the-class-name-or-cls-param-to-construct-a-an-instance) [two](https://stackoverflow.com/questions/57025928/in-a-class-method-what-is-the-difference-between-using-variable-from-classname)

JeRusakow replied:
Пользуюсь PyCharm'ом, он не возмущался на такое)
И использованием имени класса внутри метода разобрался и исправил

RomanTsaregorodtsev commented on homework6/task2/oop_2.py:
```diff
@@ -0,0 +1,145 @@
+        else:
+            return False
+
+    @staticmethod
```
Тут явно должен быть не статик метод

JeRusakow replied:
Изменил на `@classmethod`

RomanTsaregorodtsev commented on tests/test_homework6/test_oop_2.py:
```diff
@@ -0,0 +1,142 @@
+import datetime
+from unittest.mock import patch
+
+from pytest import fixture, raises
```
Обычно просто импортирую `pytest` и дальше уже везде пишут
`pytest.fixture(), pytest.raises` и так далее. По идее так лучше читается и понятно к чему именно относится значение

JeRusakow replied:
Переправил импорты

RomanTsaregorodtsev commented on tests/test_homework6/test_oop_2.py:
```diff
@@ -0,0 +1,142 @@
+
+
+@fixture
+def poor_homework_result():
```
Если ты делаешь фикстуры, то ты должен передавать их в качестве аргументов как в другие фикстуры, так и в сами тесты.

JeRusakow replied:
Исправил. Неиспользуемые фикстуры удалил

JeRusakow committed:
Changed pre-commit-config.yaml in favor of one from epam course repo

JeRusakow committed:
requirements updated to correspond the epam sample repo

JeRusakow committed:
Flake8 introduction and code fixes

[X] Flake8 introduced into project
-- Counter
[X] Counter.py does not longer uses setattr()
[X] get_created_instances and reset_instances_counter are now @classmethods
[X] getattr replaced with simple assignment
-- OOP_2
[X] Docstrings written
[X] Return typehints don't longer cause conflicts
[X] Class name is no longer used inside of class methods
-- Tests
[X] Unused fixtures deleted
[X] Used fixture names passed to tests as arguments
[X] Detailed imports e.g. `from pytest import fixture, raises` replaced with general ones: `import pytest`

JeRusakow committed:
Test filure fixed
